### PR TITLE
Copy duplicate leafs

### PR DESCRIPTION
**Pull request by weiliu89**

### CODE REVIEW

File: axlearn/common/state_builder_test.py [hunk 1]
```diff
@@ -1162,6 +1171,139 @@ def test_only_linear_weight(self):
...
+        )
+        self.assertNestedAllClose(
+            converted_state.trainer_state.model["child"]["linear"],
+            source_state.trainer_state.model["model"]["linear2"],
+        )
+        # converted_state's "linear" is donated.
+        self.assertTrue(
+            converted_state.trainer_state.model["linear"]["bias"]
+            is source_state.trainer_state.model["model"]["linear"]["bias"]
+        )
```
Comment (ruomingp) [inline axlearn/common/state_builder_test.py:1292]: Nit: use assertIs(...)?

**Review by ruomingp — APPROVED** [review]
Thanks!

File: axlearn/common/state_builder.py [hunk 2]
```diff
@@ -791,10 +792,38 @@ def target_to_source(self, target: Builder.State) -> tuple[Builder.State, Any]:
...
+                return leaf.copy()
+            copied_leaf_paths.add(path)
+            return leaf
+
         for target_scope, source_scope in self.scopes.items():
-            source_model = utils.get_recursively(source.trainer_state.model, source_scope)
+            orig_source_model = utils.get_recursively(source.trainer_state.model, source_scope)
+            source_model = jax.tree_util.tree_map_with_path(
+                lambda path, leaf: _copy_leaf(
+                    "/".join(_key_entry_to_str(k) for k in path),
```
Comment (markblee) [inline axlearn/common/state_builder.py:820]: Move this into `_copy_leaf` so we can use separator for the join?

File: axlearn/common/state_builder.py [hunk 3]
```diff
@@ -791,10 +793,41 @@ def target_to_source(self, target: Builder.State) -> tuple[Builder.State, Any]:
...
+            return leaf
+
         for target_scope, source_scope in self.scopes.items():
-            source_model = utils.get_recursively(source.trainer_state.model, source_scope)
+            orig_source_model = utils.get_recursively(source.trainer_state.model, source_scope)
+            source_model = jax.tree_util.tree_map_with_path(
+                lambda path, leaf: _copy_leaf(
+                    path,
+                    leaf,
+                    source_scope=source_scope,  # pylint: disable=cell-var-from-loop
```
Comment (markblee) [inline axlearn/common/state_builder.py:826]:
```suggestion
                lambda path, leaf, src_scope=source_scope: _copy_leaf(
                    path,
                    leaf,
                    source_scope=src_scope,
```

**Pull request merged**